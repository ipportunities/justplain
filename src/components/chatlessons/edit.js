import React, { useEffect, useState } from "react";
import { useSelector, useDispatch } from "react-redux";
import LeftMenu from "../dashboard/leftmenu";
import ContentEdit from "../content/edit";
import apiCall from "../api";
import ErrorPage from "../error";
import { setIntervention } from "../../actions";
import LoadScreen from "../loadScreen";
import t from "../translate";
import { useLocation } from "react-router-dom";

const HomeworkEdit = props => {
  let location = useLocation();
  const [contentIndex, setContentIndex] = useState(0)
  const [action, setAction] = useState(false);
  const [content, setContent] = useState("");
  const [allowed, setAllowed] = useState("loading");

  const chatlesson_id = location.pathname.split("/")[3];
  const intervention_id = useSelector(state => state.intervention.id);
  const intervention_title = useSelector(state => state.intervention.title);
  const [editable, setEditability] = useState("false");
  const [message, setMessage] = useState("");
  const [errorMessage, setErrorMessage] = useState("");
  const auth = useSelector(state => state.auth);
  const dispatch = useDispatch();

  const options = [
    "quote",
    "list",
    "wysiwyg",
    //"question_checkboxes",
    //"question_radio",
    //"question_open",
    "video",
    "audio",
    "image",
    //"select",
    "special",
    //"divider",
    "chart",
    //"feedback",
    "matrix",
    "slider",
    "datepicker",
    "custom",
  ];

  //////////////////////
  ///On init
  useEffect(() => {
    checkIfAllowedToEdit();
    if(props.action != ""){
      setAction(props.action);
    }
  }, []);

  //////////////////////
  ///Check if its allowed to edit
  function checkIfAllowedToEdit() {
    let apiMsg = {
      action: "get_chatlesson",
      token: auth.token,
      data: {
        id: chatlesson_id
      }
    };

    apiCall(apiMsg).then(resp => {
      if (resp) {
        dispatch(
          setIntervention(
            resp.intervention_id,
            resp.organisation_id,
            resp.intervention_title,
            resp.intervention_settings
          )
        );
        if (Object.keys(resp.settings).length != 0) {
          /// is de leeg dan niet
          resp.settings.newPart = false;
          resp.settings.removePart = false;
          setContent(resp.settings);
          let this_item_obj = resp.intervention_settings.chatlessons.filter(function (item) {
            return item.id === chatlesson_id
          });
          setContentIndex(resp.intervention_settings.chatlessons.indexOf(this_item_obj[0]))
        }
        setEditability(true);
        setAllowed(true);
      } else {
        setAllowed(false);
      }
    });
  }

  return (
    <div>
      <LeftMenu />
      {editable == true ? (
        <ContentEdit
          id={chatlesson_id}
          options={options}
          content={content}
          contentIndex={contentIndex}
          type="chatlesson"
          name={t("Chat les")}
          saveAction="save_chatlesson"
          action={action}
          url={"chatlessons"}
        />
      ) : (
        <div>
          {allowed == "loading" ? <div><LoadScreen/></div>:<div>
          <ErrorPage/>
          </div>}
        </div>
      )}
    </div>
  );
};

export default HomeworkEdit;
