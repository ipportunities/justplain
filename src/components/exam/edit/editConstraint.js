import React from 'react'
import { useSelector } from 'react-redux'
import apiCall from '../../api'

export const EditConstraint = ({ categoryId, subject, difficulty, amount, setAmount, updateMaxPointRange }) => {
  const [maxAmountOfQuestions, setMaxAmountOfQuestions] = React.useState(0)
  const auth = useSelector(state => state.auth)
  const [maxPointsRange, setMaxPointsRange] = React.useState([0, 0])

  React.useEffect(() => {
    apiCall({
      action: 'get_max_amount_of_questions_for_constraint',
      token: auth.token,
      data: { categoryId: categoryId, subject: subject, difficulty: difficulty }
    }).then(resp => {
      if (+resp.maxAmountOfQuestions < amount) {
        setAmount(+resp.maxAmountOfQuestions)
      }
      setMaxAmountOfQuestions(resp.maxAmountOfQuestions)
    })
  }, [subject, difficulty])

  React.useEffect(() => {
    if (amount > 0) {
      apiCall({
        action: 'get_max_points_range',
        token: auth.token,
        data: { categoryId: categoryId, subject: subject, difficulty: difficulty, amountOfQuestions: amount }
      }).then(resp => {
        updateMaxPointRange(`${categoryId}-${subject}-${difficulty}`, resp.maxPointsRange)
        setMaxPointsRange(resp.maxPointsRange)
      })
    }
  }, [subject, difficulty, amount])

  const setNewAmount = (value) => {
    setAmount(value <= maxAmountOfQuestions ? value : maxAmountOfQuestions)
  }

  return <div className="row m-0">
    <input type="number" className="form-control col-4 pl-2 pr-2" id="AmountOfQuestionsRange"
           onChange={ e => setNewAmount(+e.target.value) } min={ 0 } max={ maxAmountOfQuestions }
           name="amountOfQuestions" value={ amount } disabled={ +maxAmountOfQuestions === 0 } />
    <div className="col">{ `/ ${ maxAmountOfQuestions } (${ maxPointsRange[0] }-${ maxPointsRange[1] } pnts)` }</div>
  </div>
}