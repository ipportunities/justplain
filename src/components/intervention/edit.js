import React, { useEffect, useState } from 'react'
import { useDispatch, useSelector } from 'react-redux'
import { useLocation } from 'react-router-dom' //laten staan ivm nav in leftmenu...
import LeftMenu from '../leftmenu'
import NavBar from './settingparts/navbar.js'
import Intro from './settingparts/intro.js'
import InterventionSettingsLanguage from './settingparts/settings_language.js'
import InterventionSettingsTranslations from '../translations/translations.js'
import InterventionSettingsTitle from './settingparts/title.js'
import InterventionSettingsSubTitle from './settingparts/subtitle.js'
import InterventionSettingsApply from './settingparts/apply.js'
import InterventionSettingsType from './settingparts/type.js'
import InterventionSettingsCoach from './settingparts/coach.js'
import InterventionSettingsCourseIntro from './settingparts/courseintro.js'
import InterventionSettingsJournal from './settingparts/journal.js'
import InterventionSettingsStress from './settingparts/stress.js'
import InterventionSettingsLessons from './settingparts/lessons.js'
import InterventionSettingsChatlesson from './settingparts/chatlessons.js'
import InterventionSettingsHomework from './settingparts/homework.js'
import InterventionSettingsOptionalLessons from './settingparts/optional_lessons.js'
import InterventionSettingsGoals from './settingparts/goals.js'
import InterventionSettingsPages from './settingparts/pages.js'
import InterventionSettingsQuestionnaires from './settingparts/questionnaires.js'
import InterventionSettingsGamification from './settingparts/gamification.js'
import InterventionSettingsJson from './settingparts/json.js'
import InterventionTheme from './settingparts/thema.js'
import apiCall from '../api'
import t from '../translate'
import { setIntervention, setSavingStatus } from '../../actions'
import Saved from "./saved";
import { InterventionSettingsExamType } from './settingparts/examType'
import {appSettings} from "../../custom/settings";

let saveSettingsTimeout = null

const InterventionEdit = () => {

  let location = useLocation()
  const dispatch = useDispatch()

  const settingsType = location.pathname.split('/')[4]

  const [message, setMessage] = useState('')
  const [errorMessage, setErrorMessage] = useState('')
  const [activeTab, setActiveTab] = useState('general')

  if (settingsType !== activeTab && typeof settingsType != 'undefined' && settingsType !== '') {
    setActiveTab(settingsType)
  }

  const auth = useSelector(state => state.auth);
  const intervention = useSelector(state => state.intervention);

  const getSettings = intervention_id => {
    //api aanroepen
    apiCall({
      action: "get_intervention_settings",
      token: auth.token,
      data: {
        intervention_id
      }
    }).then(resp => {
      //om bestaande interventies te voorzien van pages...
      if (!resp.settings.hasOwnProperty("pages"))
      {
        resp.settings.pages = [];
      }
      //set settings in global state:
      dispatch(
        setIntervention(
          resp.intervention_id,
          resp.organisation_id,
          resp.title,
          resp.settings
        )
      );
    });
  };

  useEffect(() => {
    if (intervention.id == 0 || intervention.id != location.pathname.split("/")[3]) {
      let intervention_id = location.pathname.split("/")[3];
      getSettings(intervention_id);
    }
    setActiveTab("general");
  }, []);



  const cleanSettings = settings => {
    //het intervention object wordt misbruikt in de app om allerlei data aan te hangen
    //die voor het saven in de DB niet relevant is. Daarom hier een cleanup
    if (settings.hasOwnProperty("goals")) {
      for (var key in settings.goals) {
        if (settings.goals[key].hasOwnProperty("settings")) {
          delete settings.goals[key].settings;
        }
      }
    }
    if (settings.hasOwnProperty("homework")) {
      for (var key in settings.homework) {
        if (settings.homework[key].hasOwnProperty("settings")) {
          delete settings.homework[key].settings;
        }
      }
    }
    if (settings.hasOwnProperty("chatlessons")) {
      for (var key in settings.chatlessons) {
        if (settings.chatlessons[key].hasOwnProperty("settings")) {
          delete settings.chatlessons[key].settings;
        }
      }
    }
    if (settings.hasOwnProperty("pages")) {
      for (var key in settings.pages) {
        if (settings.pages[key].hasOwnProperty("settings")) {
          delete settings.pages[key].settings;
        }
      }
    }
    if (settings.hasOwnProperty("questionnaires")) {
      for (var key in settings.questionnaires) {
        if (settings.questionnaires[key].hasOwnProperty("settings")) {
          delete settings.questionnaires[key].settings;
        }
      }
    }
    if (
      settings.hasOwnProperty("selfhelp") &&
      settings.selfhelp.hasOwnProperty("lessons")
    ) {
      for (var key in settings.selfhelp.lessons) {
        if (settings.selfhelp.lessons[key].hasOwnProperty("settings")) {
          delete settings.selfhelp.lessons[key].settings;
        }
      }
      for (var key in settings.selfhelp.optionalLessons) {
        if (settings.selfhelp.optionalLessons[key].hasOwnProperty("settings")) {
          delete settings.selfhelp.optionalLessons[key].settings;
        }
      }
    }
    return settings;
  };

  const saveSettings = (intervention_id, organisation_id, title, settings) => {
    settings = cleanSettings(settings);

    apiCall({
      action: "save_intervention",
      token: auth.token,
      data: {
        intervention: {
          id: intervention_id,
          title: title,
          settings: settings,
        }
      }
    }).then(resp => {
      dispatch(setSavingStatus("saved"));
      //set settings in global state:
      dispatch(
        setIntervention(
          intervention.id,
          organisation_id,
          resp.title,
          resp.settings /// kan dit of moet de save gewoon alles teruggeven <= dat wordt ook complex .... hier zijn wel de nieuwe toegevoegd items nodig met juiste settings + deleted items....
        )
      );
      setMessage(resp.msg);
      setTimeout(() => {
        setMessage("");
      }, 5000);
    });
  };

  const showTab = tab => {
    setActiveTab(tab);
  };

  return (
    <div className="whiteWrapper intervention_edit">
      <Saved />
      <LeftMenu />
      <div className="container dashboard_container">
        <div className="settings_container">
          <InterventionSettingsTitle
            saveSettings={saveSettings}
            setErrorMessage={setErrorMessage}
          />
          <InterventionSettingsSubTitle
            saveSettings={saveSettings}
            setErrorMessage={setErrorMessage}
          />

          <NavBar showTab={showTab} />
          <br />

          <div
            className={
              errorMessage.length < 1 ? "hidden" : "alert alert-danger"
            }
            role="alert"
          >
            {errorMessage}
          </div>
            {activeTab === "general" ?
              <div id="settings-general">
                <InterventionSettingsLanguage
                  saveSettings={saveSettings}
                  setErrorMessage={setErrorMessage}
                />
                <InterventionSettingsType
                  saveSettings={saveSettings}
                  setErrorMessage={setErrorMessage}
                />
                {appSettings.introCourseField ?
                  <InterventionSettingsCourseIntro
                    saveSettings={saveSettings}
                    setErrorMessage={setErrorMessage}
                  />
                :<></>}
                <InterventionSettingsCoach
                  saveSettings={saveSettings}
                  setErrorMessage={setErrorMessage}
                />
                <InterventionSettingsJournal
                  saveSettings={saveSettings}
                  setErrorMessage={setErrorMessage}
                />
                {appSettings.stressmeter == true ?
                  <InterventionSettingsStress
                    saveSettings={saveSettings}
                    setErrorMessage={setErrorMessage}
                  />
                  :<></>}
                {appSettings == true ?
                  <InterventionSettingsApply
                    saveSettings={saveSettings}
                    setErrorMessage={setErrorMessage}
                  />
                  :<></>}
              </div>
             :<></>}
            {activeTab === "lessons" ?
              <div id="settings-lessons">
                <Intro
                  saveSettings={saveSettings}
                  setErrorMessage={setErrorMessage}
                  placeholder={t("Intro lessen")}
                  type="lessons"
                />
                <InterventionSettingsLessons
                  intervention={intervention}
                  saveSettings={saveSettings}
                  setErrorMessage={setErrorMessage}
                />

              </div>
           :<></>}

          {activeTab === "optional-lessons" ?
            <div id="settings-optional-lessons">
              <Intro
                saveSettings={saveSettings}
                setErrorMessage={setErrorMessage}
                placeholder={t("Intro optionele lessen")}
                type="optional-lessons"
              />
              <InterventionSettingsOptionalLessons
                intervention={intervention}
                saveSettings={saveSettings}
                setErrorMessage={setErrorMessage}
              />

            </div>
           :<></>}

        {activeTab === "homework" ?
          <div id="settings-homework">
            <Intro
              saveSettings={saveSettings}
              setErrorMessage={setErrorMessage}
              placeholder={t("Intro huiswerk")}
              type="homework"
            />
            <InterventionSettingsHomework
              intervention={intervention}
              saveSettings={saveSettings}
              setErrorMessage={setErrorMessage}
            />
          </div>
         :<></>}
        {activeTab === "chatlessons" ?
          <div id="settings-chatlessons">
            <InterventionSettingsChatlesson
              intervention={intervention}
              saveSettings={saveSettings}
              setErrorMessage={setErrorMessage}
            />
          </div>
         :<></>}
         {activeTab === "goals" ?
           <div id="settings-goals">
             <Intro
               saveSettings={saveSettings}
               setErrorMessage={setErrorMessage}
               placeholder="Intro doelen"
               type="goals"
             />
             <InterventionSettingsGoals
               intervention={intervention}
               saveSettings={saveSettings}
               setErrorMessage={setErrorMessage}
             />
           </div>
         :<></>}
         {activeTab === "questionnaires" ?
           <div
             id="settings-questionnaires"
           >
             <InterventionSettingsQuestionnaires
               intervention={intervention}
               saveSettings={saveSettings}
               setErrorMessage={setErrorMessage}
             />
           </div>
          : <></>}
          {activeTab === "pages" ?
            <div
              id="settings-pages"

            >
              <InterventionSettingsPages
                intervention={intervention}
                saveSettings={saveSettings}
                setErrorMessage={setErrorMessage}
              />
            </div>
           : <></>}
           {activeTab === "translations" ?
             <div
               id="settings-translations"
             >
               <InterventionSettingsTranslations
                 intervention={intervention}
                 saveSettings={saveSettings}
                 setErrorMessage={setErrorMessage}
               />
             </div>
            : <></>}
            { activeTab === 'exam-types' && appSettings.exam ?
              <div id="settings-exam-types">
                <InterventionSettingsExamType />
              </div>
              :<></>}
            { activeTab === 'gamification' && appSettings.gamification ?
              <InterventionSettingsGamification
                setErrorMessage={setErrorMessage}
                saveSettings={saveSettings}
              />
              :<></>}
            { activeTab === 'json' ?
              <div
                id="settings-json"
                className={activeTab === "json" ? "" : "hidden"}
              >
                <InterventionSettingsJson
                  saveSettings={saveSettings}
                  setErrorMessage={setErrorMessage}
                />
              </div>
              :<></>}

            {intervention.id > 0 ? (
              <div
                id="settings-theme"
                className={activeTab === "theme" ? "" : "hidden"}
              >
                <InterventionTheme
                  saveSettings={saveSettings}
                  setErrorMessage={setErrorMessage}
                />
              </div>
            ) : (
              ""
            )}
        </div>
      </div>
    </div>
  );
};

export default InterventionEdit;
