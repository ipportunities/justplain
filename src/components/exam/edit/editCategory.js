import React from 'react'
import { useSelector, useDispatch } from 'react-redux'
import { EditConstraint } from './editConstraint'
import apiCall from '../../api'
import { useTranslation } from '../../translate'
import Saved from "../../intervention/saved";
import { setSavingStatus } from "../../../actions";

export const EditExamTypeCategory = ({ examTypeId, category, examTypeCategory, setExamTypeCategory, updateQuestionsTotal, updateMaxPointRange }) => {
  const [enabled, setEnabled] = React.useState(false)
  const [saving, setSaving] = React.useState(false)
  const [saved, setSaved] = React.useState(false)
  const [id, setId] = React.useState(-1)
  const [amountOfQuestions, setAmountOfQuestions] = React.useState({})
  const dispatch = useDispatch();

  const t = useTranslation()
  const auth = useSelector(state => state.auth)

  React.useEffect(() => {
    setEnabled(examTypeCategory !== undefined)
    let newAmountOfQuestions = {}
    if (examTypeCategory !== undefined) {
      category.subjects.forEach(s => {
        newAmountOfQuestions[s] = {}
        category.difficulties.forEach(d => {
          newAmountOfQuestions[s][d] = examTypeCategory.constraints.find(c => c.subject === s && c.difficulty === d)
            ? +examTypeCategory.constraints.find(c => c.subject === s && c.difficulty === d).amountOfQuestions
            : 0
        })
      })
      setId(examTypeCategory.id)
    } else {
      let difficulties = {}
      category.difficulties.forEach(d => difficulties[d] = 0)
      category.subjects.forEach(s => newAmountOfQuestions[s] = difficulties)
    }
    setAmountOfQuestions(newAmountOfQuestions)
  }, [examTypeCategory, category])

  const switchExamTypeCategoryEnabled = () => {
    const newEnabled = !enabled
    if (newEnabled) {
      apiCall({
        action: 'add_exam_category',
        token: auth.token,
        data: { examTypeId: examTypeId, categoryId: category.id, amountOfQuestions: 0 }
      }).then(resp => {
        setExamTypeCategory(resp.examTypeCategory)
        setId(resp.examTypeCategory.id)
      })
    } else {
      apiCall({
        action: 'remove_exam_category',
        token: auth.token,
        data: { id: id }
      }).then(() => {
        setId(-1)
      })
    }
    setEnabled(newEnabled)
  }

  const saveCategory = () => {
    if (enabled) {
      let constraints = []
      Object.entries(amountOfQuestions).forEach(([subject, a]) => {
        Object.entries(a).forEach(([difficulty, amount]) => {
          constraints.push({
            subject: subject,
            difficulty: difficulty,
            amountOfQuestions: amount
          })
        })
      })
      apiCall({
        action: 'update_exam_category',
        token: auth.token,
        data: {
          examTypeId: +examTypeCategory.examTypeId,
          categoryId: +examTypeCategory.categoryId,
          amountOfQuestions: 0,
          constraints: constraints
        }
      }).then(resp => {
        dispatch(setSavingStatus("saved"));
      })
    }
  }

  React.useEffect(() => {
    dispatch(setSavingStatus("not_saved"));
    saveCategory()
    updateQuestionsTotal(Object.keys(amountOfQuestions).reduce(
      (total, s) => total + Object.keys(amountOfQuestions[s]).reduce(
        (subTotal, d) => subTotal + parseFloat(amountOfQuestions[s][d] || 0), 0
      ), 0)
    )
  }, [amountOfQuestions])

  const updateAmountOfQuestions = (subject, difficulty, amount) => {
    setAmountOfQuestions({
      ...amountOfQuestions,
      [subject]: {
        ...amountOfQuestions[subject],
        [difficulty]: +amount
      }
    })
  }

  return (
    <>
      <div className="row">
        <div className="col-10 pt-1">
          <input type="checkbox" className="form-check-input" checked={ enabled }
                 onChange={ switchExamTypeCategoryEnabled } id={ `examTypeCategoryEnabled-${ category.id }` } />
          <label className="form-check-label"
                 htmlFor={ `examTypeCategoryEnabled-${ category.id }` }>{ category.name }</label>
        </div>
        <div className="col-2 pt-1 text-right align-self-center">
          { enabled ?
            saving ?
              <div className="spinner-border text-primary" />
              : saved ?
              <div className="fa fa-2x fa-check text-success" />
              : <div />
            : <div />
          }
        </div>
      </div>
      { enabled && <>
        <table className="table">
          <thead>
          <tr>
            <th />
            { category.difficulties.map(d => <th
              key={ d }>{ d.charAt(0).toUpperCase() + d.slice(1).toLowerCase() }</th>) }
            <th className="border-left">{ t('Total') }</th>
          </tr>
          </thead>
          <tbody>
          { category.subjects.map(s => <tr className="border-top" key={ s }>
            <th>{ s === '' ? 'No Subject' : s }</th>
            { category.difficulties.map(d => <td key={ `${ d }-${ s }` }>
              <EditConstraint categoryId={ category.id } difficulty={ d } subject={ s } amount={ amountOfQuestions[s][d] }
                              setAmount={ (amount) => updateAmountOfQuestions(s, d, amount) }
                              updateMaxPointRange={ updateMaxPointRange } />
            </td>) }
            <td className="border-left">
              { Object.keys(amountOfQuestions[s]).reduce((total, i) => total + parseFloat(amountOfQuestions[s][i] || 0), 0) }
            </td>
          </tr>) }
          <tr className="border-top">
            <th>
              { t('Total') }
            </th>
            { category.difficulties.map(d => <td key={ `${ d }-total` }>{
              `${ Object.keys(amountOfQuestions).reduce((total, i) => total + parseFloat(amountOfQuestions[i][d] || 0), 0) }`
            }</td>) }
            <td className="border-left">{
              Object.keys(amountOfQuestions).reduce(
                (total, s) => total + Object.keys(amountOfQuestions[s]).reduce(
                  (subTotal, d) => subTotal + parseFloat(amountOfQuestions[s][d] || 0), 0
                ), 0)
            }</td>
          </tr>
          </tbody>
        </table>
      </> }
    </>
  )
}
