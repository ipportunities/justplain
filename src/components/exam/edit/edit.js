import React from 'react'
import { useSelector, useDispatch } from 'react-redux'
import { useHistory, useParams } from 'react-router-dom'
import { EditExamTypeCategory } from './editCategory'
import { useTranslation } from '../../translate'
import apiCall from '../../api'
import { Editor } from '@tinymce/tinymce-react'
import Saved from "../../intervention/saved";
import { setSavingStatus } from "../../../actions";

export const ExamTypeEdit = () => {
  const t = useTranslation()
  const history = useHistory()
  const { id } = useParams()
  const auth = useSelector(state => state.auth)
  const intervention = useSelector(state => state.intervention)

  const [examType, setExamType] = React.useState({})
  const [categories, setCategories] = React.useState([])
  const [examTitleTimeout, setExamTitleTimeout] = React.useState(-1)
  const [questionTotal, setQuestionTotal] = React.useState([])
  const [maxPointsRangeTotal, setMaxPointsRangeTotal] = React.useState({})
  const dispatch = useDispatch();

  const examTitleRef = React.useRef()
  examTitleRef.current = examType.title
  const examIntroTextRef = React.useRef()
  examIntroTextRef.current = examType.introText
  const examFinishTextRef = React.useRef()
  examFinishTextRef.current = examType.finishText

  const getExamType = () => {
    apiCall({
      action: 'get_exam_type',
      token: auth.token,
      data: { id: id }
    }).then(resp => {
      setExamType(resp.examType)
    })
  }

  const getCategories = () => {
    apiCall({
      action: 'get_question_categories',
      token: auth.token,
      data: { withSubjects: true, withDifficulties: true }
    }).then(resp => {
      setCategories(resp.categories)
    })
  }

  React.useEffect(() => {
    if (id >= 0) {
      getExamType()
      getCategories()
    }
  }, [id])

  const setExamData = e => {
    setExamType({ ...examType, [e.target.name]: e.target.value })
    clearTimeout(examTitleTimeout)
    dispatch(setSavingStatus("not_saved"));
    setExamTitleTimeout(setTimeout(saveExamType, 500))
  }
  const saveExamType = () => {
    apiCall({
      action: 'update_exam_type',
      token: auth.token,
      data: {
        id: id,
        title: examTitleRef.current,
        introText: examIntroTextRef.current,
        finishText: examFinishTextRef.current
      }
    }).then(resp => {
      dispatch(setSavingStatus("saved"));
      setExamType({ ...examType, ...resp.examType })
    })
  }

  const updateTotal = (index, amount) => {
    let newQuestionTotal = questionTotal
    newQuestionTotal[index] = +amount
    setQuestionTotal([...newQuestionTotal])
  }

  const updateMaxPointRange = (index, range) => {
    let newMaxPointsRangeTotal = maxPointsRangeTotal
    newMaxPointsRangeTotal[index] = range
    setMaxPointsRangeTotal({ ...newMaxPointsRangeTotal })
  }

  return (
    <div className="whiteWrapper">
      <div className="lessoncontent exam edit">
        <nav className="navbar navbar-expand-lg navbar-light overRule">
          <h2>
            <span className="pointer" onClick={ () => history.push(`/intervention/edit/${ intervention.id }/general/`) }>
                { `${ intervention.title } > ` }
            </span>
            <span className="pointer"
                  onClick={ () => history.push(`/intervention/edit/${ intervention.id }/exam-types/`) }>
              { t('Examens') }
            </span>
          </h2>
        </nav>
        <Saved/>
        <div className="settings_container">
          <h1>{ t('Wijzig examen') }</h1>
          <div className="row pb-1 pt-5">
            <div className="col-6">
              <div className="form-group">
                <label htmlFor="categoryName">{ t('Titel') }</label>
                <input type="text" className="form-control pl-2 pr-2" id="examTitle"
                       value={ examType.title } name="title" onChange={ setExamData } />
              </div>
            </div>
            <div className="col-4 align-self-center bleu">
              {t("Vragen")}{ ` :  ${ questionTotal.reduce((a, b) => a + b, 0) }` }
              <br />
              {t("Punten")}{ ` :  ${ Object.values(maxPointsRangeTotal).reduce((a, b) => a + b[0], 0) } - ${ Object.values(maxPointsRangeTotal).reduce((a, b) => a + b[1], 0) } ` }
            </div>
          </div>
          <div className="row pb-2">
            <div className="col">
              <div className="form-group">
                <label htmlFor="introText">{ t('Intro') }</label>
                <Editor
                  apiKey="k68mc81xjxepc3s70sz7ns6ddgsx6bcgzpn3xgftlxgshmb3"
                  inline
                  value={ examType.introText }
                  init={ {
                    menubar:false,
                    plugins: 'link image code lists advlist',
                    toolbar: [
                      "undo redo | styleselect | bold italic | alignleft aligncenter alignright alignjustify",
                      "bullist numlist outdent indent | table | fontsizeselect | link "],
                  } }
                  onEditorChange={ (content, editor) => setExamData({ target: { name: 'introText', value: content } }) }
                />
              </div>
            </div>
          </div>
          <div className="row pb-2">
            <div className="col">
              <div className="form-group">
                <label htmlFor="finishText">{ t('Afsluiting') }</label>
                <Editor
                  apiKey="k68mc81xjxepc3s70sz7ns6ddgsx6bcgzpn3xgftlxgshmb3"
                  inline
                  value={ examType.finishText }
                  init={ {
                    menubar:false,
                    plugins: 'link image code lists advlist',
                    toolbar: [
                      "undo redo | styleselect | bold italic | alignleft aligncenter alignright alignjustify",
                      "bullist numlist outdent indent | table | fontsizeselect | link "],
                  } }
                  onEditorChange={ (content, editor) => setExamData({
                    target: {
                      name: 'finishText',
                      value: content
                    }
                  }) }
                />
              </div>
            </div>
          </div>
          { examType.examTypeCategories !== undefined ? <>
            <label>{ t('Onderdelen') }</label>
            { categories.filter(c => +c.amountOfQuestions > 0).sort((a, b) => a.id - b.id).map((category, index) => {
              return <div className={ `settings_container examtype_category` }
                          key={ category.id }>
                <EditExamTypeCategory
                  category={ category }
                  examTypeId={ examType.id }
                  updateMaxPointRange={ updateMaxPointRange }
                  updateQuestionsTotal={ (amount) => updateTotal(index, amount) }
                  examTypeCategory={ examType.examTypeCategories.find(ec => ec.categoryId === category.id) }
                  setExamTypeCategory={ examTypeCategory => setExamType({
                    ...examType,
                    examTypeCategories: [...examType.examTypeCategories, examTypeCategory]
                  }) } />
              </div>
            }) }
          </> : null }
        </div>
      </div>
    </div>
  )
}
