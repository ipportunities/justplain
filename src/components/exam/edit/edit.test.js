import React from 'react'
import { shallow } from 'enzyme'
import { ExamTypeEdit } from './edit'

jest.mock('react-redux', () => ({
  ...jest.requireActual('react-redux'),
  useSelector: jest.fn().mockReturnValue({ auth: { token: 'a' }, intervention: { id: 1, title: 'some intervention' } })
}))

jest.mock('react-router-dom', () => ({
  ...jest.requireActual('react-router-dom'), // use actual for all non-hook parts
  useParams: () => ({
    id: 0,
  }),
  useHistory: () => ({
    push: jest.fn()
  })
}))

describe('Exam Type Edit', () => {
  it('should render correctly', () => {
    const component = shallow(<ExamTypeEdit />)
    expect(component).toMatchSnapshot()
  })
})
