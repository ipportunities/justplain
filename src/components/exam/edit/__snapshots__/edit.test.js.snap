// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Exam Type Edit should render correctly 1`] = `
<div
  className="lessoncontent edit"
>
  <nav
    className="navbar navbar-expand-lg navbar-light overRule"
  >
    <h2>
      <span
        className="pointer"
        onClick={[Function]}
      >
        undefined &gt; 
      </span>
      <span
        className="pointer"
        onClick={[Function]}
      >
        Exam Type
      </span>
    </h2>
  </nav>
  <div
    className="row border-bottom border-info"
  >
    <div
      className="col-6 offset-3"
    >
      <div
        className="row"
      >
        <h1>
          Edit exam type
        </h1>
      </div>
      <div
        className="row pb-1 pt-5"
      >
        <form
          className="col-8"
        >
          <div
            className="form-group"
          >
            <label
              htmlFor="categoryName"
            >
              Exam title
            </label>
            <input
              className="form-control pl-2 pr-2"
              id="examTitle"
              name="title"
              onChange={[Function]}
              type="text"
            />
          </div>
        </form>
        <div
          className="col-4 align-self-center"
        >
          Total amount of questions: 0
          <br />
          Maximum achievable points: 0 - 0 pnts
        </div>
      </div>
      <div
        className="row pb-2"
      >
        <div
          className="col"
        >
          <div
            className="form-group"
          >
            <label
              htmlFor="introText"
            >
              intro text
            </label>
            <input
              className="form-control"
              id="introText"
              name="introText"
              onChange={[Function]}
              type="text"
            />
          </div>
        </div>
      </div>
      <div
        className="row pb-2"
      >
        <div
          className="col"
        >
          <div
            className="form-group"
          >
            <label
              htmlFor="finishText"
            >
              finish text
            </label>
            <input
              className="form-control"
              id="finishText"
              name="finishText"
              onChange={[Function]}
              type="text"
            />
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
